File: src/GUI/objects/PopOver/Notification.java

package GUI.objects.PopOver;

import GUI.objects.PopOver.PopOver.ScreenCentricPos;
import javafx.scene.Node;
import static javafx.scene.input.MouseButton.PRIMARY;
import static javafx.scene.input.MouseButton.SECONDARY;
import javafx.scene.layout.StackPane;
import javafx.util.Duration;
import util.async.FxTimer;

/**
 * Notification popover.
 */
public class Notification extends PopOver {
        
    private final FxTimer closer = new FxTimer(5000, 1, this::hide);
    private final StackPane root = new StackPane();
    
    /** Executes on left mouse click. Default null. */
    public Runnable lClickAction = null;
    /** Executes on right mouse click. Default null. */
    public Runnable rClickAction = null;
    
    public Notification() {
        detached.set(false);
        detachable.set(false);
        setHideOnEscape(false);
        setArrowSize(0);
        setArrowIndent(0);
        setCornerRadius(0);
        setAutoFix(false);
        setAutoHide(false);
        getSkinn().setTitleAsOnlyHeaderContent();
        getSkinn().root.getStyleClass().add("notification");
        
        setContentNode(root);
        root.setOnMouseClicked(e -> {
            if(e.getButton()==PRIMARY && lClickAction!=null ) lClickAction.run();
            if(e.getButton()==SECONDARY && rClickAction!=null ) rClickAction.run();
        });
        root.setOnMouseEntered(e-> closer.pause());
        root.setOnMouseExited(e-> closer.unpause());
    }
    
    @Override
    public void show(ScreenCentricPos pos) {
        super.show(pos);
        closer.restart();
    }
    
    /**
     * Sets content of the notifier using data from provided metadata.
     * @param content of the notification. Content type depends on type parameter
     * and is not arbitrary. If type is PLAYBACK or SONG, content must be Metadata.
     * For TEXT it must be String and for OTHER Node is required.
     * @param type 
     * @param title 
     * @throws NullPointerException if param null
     * @throws ClassCastException if 
     * content not Metadata if type PLAYBACK or SONG
     * content not Node if type OTHER
     * content not String if type TEXT
     */
    public void setContent(Node n, String title) {
        headerVisible.set(!title.isEmpty());
        this.title.set(title);
        n.setMouseTransparent(true);
        setContentNode(null);
        root.getChildren().setAll(n);
        setContentNode(root);
    }
    

    /** Returns time this notification will remain visible. Default 5 seconds. */
    public Duration getDuration() {
        return closer.getPeriod();
    }

    /** Sets time this notification will remain visible. */
    public void setDuration(Duration duration) {
        closer.setPeriod(duration);
    }
}

File: src/services/notif/Notification.kt
package services.notif

import gui.objects.popover.PopOver
import gui.objects.popover.ScreenPos
import javafx.scene.Node
import javafx.scene.input.MouseButton.PRIMARY
import javafx.scene.input.MouseButton.SECONDARY
import javafx.scene.layout.StackPane
import javafx.util.Duration
import util.async.executor.FxTimer

/** Notification popover. */
class Notification: PopOver<Node>() {
    private val closer = FxTimer(5000.0, 1, Runnable { this.hide() })
    private val root = StackPane()

    /** Executes on left mouse click. Default does nothing. */
    var lClickAction = Runnable {}

    /** Executes on right mouse click. Default does nothing. */
    var rClickAction = Runnable {}

    /** Time this notification will remain visible. Default 5 seconds. */
    var duration: Duration
        get() = closer.period
        set(duration) { closer.period = duration }

    init {
        detached.value = false
        detachable.value = false
        userResizable.value = false
        isHideOnEscape = false
        arrowSize = 0.0
        arrowIndent = 0.0
        cornerRadius = 0.0
        isAutoFix = false
        isAutoHide = false
        skinn.setTitleAsOnlyHeaderContent(true)
        styleClass += "notification"

        contentNode = root
        root.setOnMouseClicked {
            if (it.button==PRIMARY) lClickAction.run()
            if (it.button==SECONDARY) rClickAction.run()
        }
        root.setOnMouseEntered { closer.pause() }
        root.setOnMouseExited { closer.unpause() }
    }

    override fun show(pos: ScreenPos) {
        super.show(pos)
        closer.start()
    }

    fun setContent(content: Node, titleText: String) {
        headerVisible.value = !titleText.isEmpty()
        title.value = titleText
        root.children.setAll(content)
        content.isMouseTransparent = true
    }

}